##  Lesson 2. Strings，Lists，Tuples


## 2.1 Strings

### 2.1.1 A string is a sequence

In [77]:
fruit = 'banana'

In [78]:
print fruit

banana


In [79]:
print fruit[0], fruit[1]

b a


### 2.1.2 len()

In [80]:
# return the length of a string
len(fruit)

6

In [81]:
len("")

0

### 2.1.3 Traversal with a for loop

In [82]:
for letter in fruit:
    print letter

b
a
n
a
n
a


In [83]:
prefix = 'JKLMNOPQ'
suffix = 'ack'

for l in prefix:
    print l + suffix

Jack
Kack
Lack
Mack
Nack
Oack
Pack
Qack


### 2.1.4 String operations

In [84]:
3 * 'Python'

'PythonPythonPython'

In [85]:
0 * "python"

''

In [86]:
'python' + 'er'

'pythoner'

In [87]:
'python'     'er'

'pythoner'

In [88]:
t = ['2012-01-01', '2010-02-01', '2010-03-01',
     '2012-04-01', '2010-05-01', '2010-06-01',
     '2012-07-01', '2010-08-01', '2010-09-01',
    ]

In [89]:
t

['2012-01-01',
 '2010-02-01',
 '2010-03-012012-04-01',
 '2010-05-01',
 '2010-06-01',
 '2012-07-01',
 '2010-08-01',
 '2010-09-01']

In [90]:
'python' 'Python'

'pythonPython'

In [91]:
prefix = 'Py'
prefix 'thon' # can't concatenate a variable and a string literal

SyntaxError: invalid syntax (<ipython-input-91-19a4deff3439>, line 2)

In [92]:
('un' * 3) 'ium'

SyntaxError: invalid syntax (<ipython-input-92-826b8aeb7d3b>, line 1)

In [93]:
('un' * 3) + 'ium'

'unununium'

In [94]:
text = ('Put several strings within parentheses '
         'to have them joined together.')

In [95]:
text

'Put several strings within parentheses to have them joined together.'

### 2.1.5 String Indices

|character|	c|	o|	m|	p|	u|	t|	e|	r|
|------|------|
|index|	0|	1|	2|	3|	4|	5|	6|	7|
|index from the right end|	-8|	-7|	-6|	-5|	-4|	-3|	-2|	-1|

In [96]:
#    01234567
s = 'computer'
print s[0]
print s[5]

c
t


Attempting to use an index that is too large will result in an error:

In [97]:
s[8]

IndexError: string index out of range

|character|	c|	o|	m|	p|	u|	t|	e|	r|
|------|------|
|index|	0|	1|	2|	3|	4|	5|	6|	7|
|index from the right end|	-8|	-7|	-6|	-5|	-4|	-3|	-2|	-1|

In [98]:
# negative index
print s[-1]
print s[-3]

r
t


In [99]:
s[-10]

IndexError: string index out of range

### 2.1.6 String slices
A segment of a string is called a slice.

string[start:end]

In [100]:
s = 'Hello Python!'

In [101]:
len(s)

13

In [103]:
s[:6]  # characters from position 0 (included) to 6 (excluded)

'Hello '

In [102]:
s[0:6]

'Hello '

In [104]:
s[6:]  # characters from position 6 (included) to 13 (excluded)

'Python!'

In [105]:
s[6:len(s)]

'Python!'

In [106]:
s[:6] + s[6:]

'Hello Python!'

In [107]:
s[:6] + s[6:] == s

True

In [39]:
s[3:7]

'lo P'

Out of range slice indexes are handled gracefully when used for slicing:

In [108]:
s[3:20], s[3:]

('lo Python!', 'lo Python!')

In [109]:
s[6:-1]

'Python'

In [110]:
s[20:]

''

In [111]:
s[-1]

'!'

In [112]:
s[:-1]

'Hello Python'

In [113]:
s[:-10]

'Hel'

In [16]:
s[-2:-1] # characters from position -2 (included) to -1 (excluded)

'n'

In [17]:
s[-3:-1]

'on'

In [114]:
s[1:1]

''

In [115]:
# What does s[:] mean?
t = s[:]

In [116]:
t

'Hello Python!'

In [118]:
assert t == s and t is s

In [119]:
id(t)

4368232320

In [120]:
id(s)

4368232320

#### Advanced: someString[start:end:stride]

In [121]:
s

'Hello Python!'

In [122]:
s[::2]

'HloPto!'

In [77]:
s[1::2]

'el yhn'

In [78]:
s[::-1]

'!nohtyP olleH'

In [44]:
s[1:10:3]

'eoy'

In [45]:
s[1:20:4]

'e h'

### 2.1.7 Strings is immutable

In [123]:
name = 'Jack'
name[0] = 'H'

TypeError: 'str' object does not support item assignment

### 2.1.8 Search, loop and count

In [124]:
# find the index of a character

index = 0
word = 'Pineapple'
while index < len(word):
    if word[index] == 'p':
        break
    index += 1
print index

5


In [132]:
word.find?

In [125]:
word.find('p')

5

In [134]:
word.find('p')

5

In [127]:
# Count the number of times the letter appears in a string

count = 0
for letter in word:
    if letter == 'e':
        count += 1
print count

2


In [128]:
word.count('e')

2

### 2.1.9 String methods

In [51]:
print word

Pineapple


In [52]:
word.find?

In [47]:
word.find('Pi')

0

In [135]:
word.find('mit')

-1

In [30]:
word.index('P')

0

In [43]:
word.find?

In [138]:
word.index?

In [136]:
word.find('W')

-1

In [166]:
word

'Pineapple'

In [167]:
word.find('i', 0, 7)

1

In [137]:
word.index('W')

ValueError: substring not found

In [139]:
word.swapcase()

'pINEAPPLE'

In [140]:
word.lower()

'pineapple'

In [141]:
word.islower()

False

In [143]:
word.lower().islower()

True

In [144]:
word.upper()

'PINEAPPLE'

In [146]:
# Swapcase
result = ''
for l in word:
    if l.islower():
        result += l.upper()
    elif l.upper():
        result += l.lower()
    else:
        result += l
print result

pINEAPPLE


In [148]:
zip_code = '975'

In [149]:
zip_code.zfill(5)

'00975'

In [150]:
zip_code.zfill(10)

'0000000975'

In [54]:
'word'.capitalize()

'Word'

In [151]:
word.center?

### 2.1.10 The in operator

In [152]:
'a' in 'banana'

True

In [153]:
'm' in 'banana'

False

### 2.1.11 String comparision

In [154]:
a = 'banana'
b = 'bananas'
c = 'apple'
d = 'Apple'

In [155]:
a < b

True

In [159]:
b < c

False

In [158]:
c < d

False

In [156]:
ord('a')

97

In [157]:
ord('A')

65

### Question: 1. Reverse a String

In [46]:
s = 'Python is good!'

In [160]:
# Method 1:
print s[::-1]

!nohtyP olleH


In [161]:
# Method 2:
t = ""
for i in s:
    t = i + t
print t

!nohtyP olleH


In [162]:
# Method 3:
"".join(reversed(list(s)))

'!nohtyP olleH'

### Question: 2. Palindrome

Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Follow up: How about not ignoring cases?

In [163]:
s = "A man, a plan, a canal: Panama"

In [80]:
s = "race a car"

In [ ]:
" .,., "

In [164]:
'1'.lower()

'1'

In [ ]:
'abcba'

In [81]:
if len(s) == 0:
    print True
else:        
    i, j = 0, len(s) - 1
    while i < j:
        while i < j and not s[i].isalnum():
            i += 1
        while i < j and not s[j].isalnum():
            j -= 1

        if s[i].lower() != s[j].lower():
            print False
            break
        else:
            i += 1
            j -= 1
    if i >= j:
        print True

False


In [165]:
import antigravity

## 2.2 Lists

### 2.2.1 A list is a sequence

In [168]:
numbers = [1, 2, 3, 4]
print numbers

[1, 2, 3, 4]


In [169]:
['crunchy frog', 'ram bladder', 'lark vomit']

['crunchy frog', 'ram bladder', 'lark vomit']

In [170]:
['spam', 2.0, 1 + 2j, 5, [1, 2]]

['spam', 2.0, (1+2j), 5, [1, 2]]

In [171]:
[]

[]

### 2.2.2 Lists are mutable

In [172]:
print numbers

[1, 2, 3, 4]


In [174]:
numbers[0] = 100
numbers[-1] = 200
print numbers

[100, 2, 3, 200]


### 2.2.3 Traversing a list

In [175]:
for num in numbers:
    print num * 2

200
4
6
400


In [176]:
for i in range(len(numbers)):
    print numbers[i]

100
2
3
200


In [177]:
for i in []:
    print 'Hello!'

### 2.2.4 List Indices and slices

In [178]:
t = ['a', 'b', 'c', 'd', 'e', 'f', 'g']

List Indices

In [179]:
t[0]

'a'

In [180]:
t[3]

'd'

In [181]:
t[-1]

'g'

In [182]:
t[-20]

IndexError: list index out of range

List slices

In [183]:
t[0:3]

['a', 'b', 'c']

In [184]:
t[-2:-1]

['f']

In [185]:
t[3:2]

[]

In [186]:
t[-1:-2]

[]

In [187]:
# If you leave out both the start and the end indexes when slicing, you'll end up with a copy of the original list
t1 = t[:]

In [189]:
t1 == t

True

In [188]:
assert t1 == t and t1 is t

AssertionError: 

In [191]:
id(t1), id(t)

(4368853256, 4368751360)

In [190]:
t

['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [192]:
t1[0] = 100

In [193]:
print t1
print t

[100, 'b', 'c', 'd', 'e', 'f', 'g']
['a', 'b', 'c', 'd', 'e', 'f', 'g']


somelist[start:end:stride]

In [131]:
t

[101, 102, 103]

In [194]:
t[::-1]

['g', 'f', 'e', 'd', 'c', 'b', 'a']

In [195]:
t[::2]

['a', 'c', 'e', 'g']

In [138]:
t

['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [196]:
t[-2:2:-2]

['f', 'd']

In [197]:
t2 = t

In [198]:
t

['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [199]:
t[0] = 100

In [200]:
t2

[100, 'b', 'c', 'd', 'e', 'f', 'g']

### 2.2.5 List methods

In [201]:
m = ['a', 'b', 'c']

In [202]:
# append
m.append('d')

In [203]:
print m

['a', 'b', 'c', 'd']


In [204]:
# extend takes a list as an argument and appends all of the elements
m1 = ['a', 'b', 'c']
m2 = ['e', 'f']
m1.extend(m2)

In [205]:
print m1

['a', 'b', 'c', 'e', 'f']


In [206]:
m1

['a', 'b', 'c', 'e', 'f']

In [207]:
# sort()
m1[0] = 'g'
m1.sort()
print m1

['b', 'c', 'e', 'f', 'g']


In [208]:
m1.sort(reverse=True)
print m1

['g', 'f', 'e', 'c', 'b']


In [212]:
# join
s = ' '.join(m1)
print s

g f e c b


In [214]:
s = '__'.join(m1)

In [215]:
s

'g__f__e__c__b'

In [216]:
# split
s.split('__')

['g', 'f', 'e', 'c', 'b']

In [218]:
text = "Python is the best programming language.\t\nI like Python!"

In [219]:
text.split('\t\n')

['Python is the best programming language.', 'I like Python!']

In [220]:
m1

['g', 'f', 'e', 'c', 'b']

In [221]:
print m1
m1.pop()
print m1

['g', 'f', 'e', 'c', 'b']
['g', 'f', 'e', 'c']


In [222]:
m1.pop(0)

'g'

In [223]:
m1

['f', 'e', 'c']

In [224]:
print m1

['f', 'e', 'c']


In [225]:
m1.remove('e')

In [226]:
print m1

['f', 'c']


In [227]:
del m1[0]

In [228]:
m1

['c']

### 2.2.6 List Comprehensions
Python provides compact syntax for deriving one list from another.

In [229]:
a = range(1, 11)
a

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [230]:
squares = []
for num in a:
    squares.append(num**2)
print squares

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [232]:
[x**2 for x in a]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [233]:
[x**2 for x in a if x % 2 == 0]

[4, 16, 36, 64, 100]

In [235]:
matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
flat = [x for row in matrix for x in row]
print flat

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [236]:
res = []
for row in matrix:
    res += row
print res

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [185]:
flat1 = []
for row in matrix:
    for x in row:
        flat1.append(x)
print flat1

[1, 2, 3, 4, 5, 6, 7, 8, 9]


### Question: Matrix Transpose

In [237]:
# Sample 1
#  1  2  3        1  4  7 10
#  4  5  6  ----> 2  5  8 11
#  7  8  9        3  6  9 12
# 10 11 12
# formula A[i][j] <---> AT[j][i]

matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]]

In [61]:
# matrix = [[]]

In [238]:
[0] * 4

[0, 0, 0, 0]

In [239]:
[[0] * 3 ] * 3

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]

In [240]:
if len(matrix) == 0 or len(matrix[0]) == 0:
    print "Result is []"
else:
    m = len(matrix)
    n = len(matrix[0])
    mt = [[0] * m for i in range(n)]

    for i in range(n):
        for j in range(m):
            mt[i][j] = matrix[j][i]
    print mt

[[1, 4, 7, 10], [2, 5, 8, 11], [3, 6, 9, 12]]


## 2.3 Tuples

### 2.3.1 Tuples are immutable

In [241]:
t = 'a', 'b', 'c'

In [242]:
t

('a', 'b', 'c')

In [243]:
t = ('a', 'b', 'c')

In [244]:
t = ("Julia", "Roberts", 1967, "Duplicity", 2009, "Actress", "Atlanta, Georgia")

A value in parentheses is not a tuple:

In [246]:
t1 = ('a')

In [248]:
t1

'a'

In [247]:
type(t1)

str

To create a tuple with a single element, we have to include the final comma

In [249]:
t2 = ('a',)

In [251]:
t2

('a',)

In [252]:
type(t2)

tuple

Create a tuple wiht built-in function

In [253]:
t = tuple()

In [254]:
type(t)

tuple

In [255]:
print t

()


In [256]:
l = list()

In [257]:
l

[]

In [258]:
t = tuple('Python')

In [259]:
print t

('P', 'y', 't', 'h', 'o', 'n')


In [260]:
t[0]

'P'

In [261]:
t[1:3]

('y', 't')

Tuplea are immutable

In [262]:
t[0] = 'p'

TypeError: 'tuple' object does not support item assignment

But we can slice parts of the old tuple and join up the bits to make the new tuple

In [263]:
t = ('p',) + t[1:]

In [264]:
print t

('p', 'y', 't', 'h', 'o', 'n')


In [265]:
l = list()
t = (l , l)

In [267]:
l.append(1)

In [268]:
t

([1, 1], [1, 1])

In [269]:
l = [2]
t = (l , l)
l.append(3)

In [270]:
t

([2, 3], [2, 3])

### 2.3.2 Tuple assignment

In [271]:
# swap a and b
a = 2
b = 100
print a, b

temp = a
a = b
b = temp
print a, b

2 100
100 2


Tuple assignment is elegent

In [272]:
a, b = b, a
# (a, b) = (b, a)

The left side is a tuple of variables; the right side is a tuple of values. Each value is assigned to its respective variable. All the expressions on the right side are evaluated before any of the assignments.

In [273]:
print a, b

2 100


The number of variables on the left and the number of values on the right have to be the same:

In [274]:
a, b = 1, 100

In [275]:
a, b = 1, 2, 3

ValueError: too many values to unpack

The right side can be any kind of sequence (string, list or tuple)

In [276]:
a, b = 'ab'
print a, b

a b


In [277]:
a, b = 'qishi@gmail.com'.split('@')

In [278]:
print a, b

qishi gmail.com


### Question. Climbing Stairs

You are climbing a stair case. It takes n steps to reach to the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

In [34]:
# s(n) = s(n-1) + s(n-2)
# Fibonacci Sequence

1,1,2,3,5,8,13,21..

In [279]:
n = 10
a, b = 1, 1
for i in range(10):
    temp = b
    b = a + b
    a = temp
print a

89


In [280]:
n = 10
a, b = 1, 1
for i in range(10):
    a, b = b, a+b
print a

89


## Summary

### 1. String

### 2. Lists

### 3. Tuple

## Assignments:

1. Reverse Vowels of a String

        Sample Input: "hello"
        
        Output: "holle"
        
2. Matrix Multiplication

        Sample Input1: A = 12 56
                           45 78
                       B = 2 6
                           5 8
        Output1: 304 520
                 480 894
                
        Sample Input2: A = 1 2 3
                           4 5 6
                           7 8 9
                       B = 1 2 3
                           4 5 6
                           7 8 9

        Output2: 30  36  42  
                 66  81  96  
                 102 126 150                
3. Reverse a List
        
        Sample Input: [1, 2, 3, 'Python']
        
        Output: ['Python', 3, 2, 1]
        
4. Given s = ['a', 'b', 'c', 'e', 'f', 'g'], try s[2::2], 2[2:2:2], s[-2::-2], s[-2:2:-2]